### PR TITLE
[Refactor]: Task Manager

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
based on #591 
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Refactor shutdown
* what changes does this pull request make?
Use `TaskManager` to manage all tasks, Handle dependencies between tasks in `TaskManager`
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: xline/src/server/watch_server.rs [hunk 1]
```diff
@@ -582,16 +587,17 @@ mod test {
         for count in collection.lock().values() {
             assert_eq!(*count, 1);
         }
-        handle1.abort();
-        handle2.abort();
-        tx.self_shutdown_and_wait().await;
+        // handle1.abort();
+        // handle2.abort();
+        // TODO
```
Comment (bsbds) [inline xline/src/server/watch_server.rs:592]: Please remove the comments

File: utils/src/task_manager/tasks.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    (TaskName::ConflictCheckedMpmc, TaskName::CmdWorker),
+    (TaskName::CmdWorker, TaskName::CompactBg),
+    (TaskName::CmdWorker, TaskName::KvUpdates),
+    (TaskName::KvUpdates, TaskName::WatchTask),
+    (TaskName::LeaseKeepAlive, TaskName::TonicServer),
+    (TaskName::TonicServer, TaskName::WatchTask),
+    (TaskName::TonicServer, TaskName::ConfChange),
+    (TaskName::TonicServer, TaskName::LogPersist),
+    (TaskName::Election, TaskName::LogPersist),
+];
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply utils/src/task_manager/tasks.rs:62]: I don't quite understand what you mean, can you provide me with an example?

File: utils/src/task_manager/mod.rs [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+/// Task names and edges
+pub mod tasks;
+
+/// Task manager
+#[derive(Debug)]
+pub struct TaskManager {
+    /// All tasks
+    tasks: Arc<DashMap<TaskName, Task>>,
+    /// State of task manager
+    state: Arc<AtomicU8>,
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply utils/src/task_manager/mod.rs:26]: `Arc<State>`requires a lock

File: utils/src/task_manager/mod.rs [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        tasks
+            .iter()
+            .filter_map(|task| (task.depend_cnt == 0).then_some(task.name))
+            .collect()
+    }
+
+    /// Inner shutdown task
+    async fn inner_shutdown(tasks: Arc<DashMap<TaskName, Task>>, state: Arc<AtomicU8>) {
+        let mut queue = Self::root_tasks_queue(&tasks);
+        state.store(1, Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (themanforfree) [reply utils/src/task_manager/mod.rs]: Done

File: curp/src/server/raw_curp/mod.rs [hunk 5]
```diff
@@ -1161,25 +1161,14 @@ impl<C: Command, RC: RoleChange> RawCurp<C, RC> {
...
-        !matches!(self.shutdown_trigger.state(), Signal::Running)
-    }
-
-    /// Get a shutdown listener
-    pub(super) fn shutdown_trigger(&self) -> &shutdown::Trigger {
-        &self.shutdown_trigger
+        self.task_manager.is_shutdown()
     }
 
     /// Get a shutdown listener
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/mod.rs:1169]: Please modify the comment.

File: utils/src/task_manager/mod.rs [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    /// Get root tasks queue
+    fn root_tasks_queue(tasks: &DashMap<TaskName, Task>) -> VecDeque<TaskName> {
+        tasks
+            .iter()
+            .filter_map(|task| (task.depend_cnt == 0).then_some(task.name))
+            .collect()
+    }
+
+    /// Inner shutdown task
+    async fn inner_shutdown(tasks: Arc<DashMap<TaskName, Task>>, state: Arc<AtomicU8>) {
```
Comment (Phoenix500526) [inline utils/src/task_manager/mod.rs:183]: Please add an unit test to cover this method, make sure that the topology order is always as you expected.

**Review by Phoenix500526 — DISMISSED** [review]
LGTM

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Review by bsbds — APPROVED** [review]

**Pull request merged**